### PR TITLE
[#3982]feat(flink-connector):Support more column data type for hive table

### PR DESCRIPTION
**Pull request by coolderli**
### What changes were proposed in this pull request?

- Support more columns for hive table

### Why are the changes needed?

Fix: #3982 

### Does this PR introduce _any_ user-facing change?

- no

### How was this patch tested?

- add UTs and ITs: flink-connector/src/test/java/org/apache/gravitino/flink/connector/integration/test/hive/FlinkHiveCatalogIT.java
- flink-connector/src/test/java/org/apache/gravitino/flink/connector/utils/TestTypeUtils.java

### CODE REVIEW

File: flink-connector/build.gradle.kts [hunk 1]
```diff
@@ -33,6 +33,7 @@ val artifactName = "gravitino-${project.name}-$scalaVersion"
 dependencies {
   implementation(project(":api"))
   implementation(project(":catalogs:catalog-common"))
+  implementation(project(":catalogs:catalog-hive"))
```
Comment (FANNG1) [inline flink-connector/build.gradle.kts:36]: why depend on `catalog-hive`?

**Comment by FANNG1** [comment]
@coolderli , could you resolve the conflict?

File: flink-connector/build.gradle.kts (hunk 1 above)
Comment (coolderli) [reply flink-connector/build.gradle.kts:36]: There is an exception.
```
FlinkHiveCatalogIT > initializationError FAILED
    java.lang.IllegalArgumentException: Failed to operate catalog(s) [test_flink_hive_schema_catalog] operation [CREATE] under metalake [flink], reason [Invalid package path: /home/mi/IdeaProjects/gravitino/catalogs/catalog-hive/build/libs in [/home/mi/IdeaProjects/gravitino/catalogs/catalog-hive/build/libs, /home/mi/IdeaProjects/gravitino/catalogs/catalog-hive/build/resources/main]]
    java.lang.IllegalArgumentException: Invalid package path: /home/mi/IdeaProjects/gravitino/catalogs/catalog-hive/build/libs in [/home/mi/IdeaProjects/gravitino/catalogs/catalog-hive/build/libs, /home/mi/IdeaProjects/gravitino/catalogs/catalog-hive/build/resources/main]
        at org.apache.gravitino.utils.IsolatedClassLoader.buildClassLoader(IsolatedClassLoader.java:121)
        at org.apache.gravitino.catalog.CatalogManager.createClassLoader(CatalogManager.java:711)
        at org.apache.gravitino.catalog.CatalogManager.createCatalogWrapper(CatalogManager.java:659)
        at org.apache.gravitino.catalog.CatalogManager.lambda$createCatalog$6(CatalogManager.java:376)
        at com.github.benmanes.caffeine.cache.BoundedLocalCache.lambda$doComputeIfAbsent$14(BoundedLocalCache.java:2406)
        at java.util.concurrent.ConcurrentHashMap.compute(ConcurrentHashMap.java:1853)
        at com.github.benmanes.caffeine.cache.BoundedLocalCache.doComputeIfAbsent(BoundedLocalCache.java:2404)
        at com.github.benmanes.caffeine.cache.BoundedLocalCache.computeIfAbsent(BoundedLocalCache.java:2387)
        at com.github.benmanes.caffeine.cache.LocalCache.computeIfAbsent(LocalCache.java:108)
        at com.github.benmanes.caffeine.cache.LocalManualCache.get(LocalManualCache.java:62)
        at org.apache.gravitino.catalog.CatalogManager.createCatalog(CatalogManager.java:376)
        at org.apache.gravitino.catalog.CatalogNormalizeDispatcher.createCatalog(CatalogNormalizeDispatcher.java:86)
        at org.apache.gravitino.listener.CatalogEventDispatcher.createCatalog(CatalogEventDispatcher.java:115)
        at org.apache.gravitino.server.web.rest.CatalogOperations.lambda$createCatalog$2(CatalogOperations.java:135)
        at org.apache.gravitino.lock.TreeLockUtils.doWithTreeLock(TreeLockUtils.java:49)
        at org.apache.gravitino.server.web.rest.CatalogOperations.lambda$createCatalog$3(CatalogOperations.java:131)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.Subject.doAs(Subject.java:422)
        at org.apache.gravitino.utils.PrincipalUtils.doAs(PrincipalUtils.java:39)
        at org.apache.gravitino.server.web.Utils.doAs(Utils.java:149)
        at org.apache.gravitino.server.web.rest.CatalogOperations.createCatalog(CatalogOperations.java:125)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:146)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:189)
        at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:93)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:478)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:400)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:81)
        at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:256)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
        at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:235)
        at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:684)
        at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394)
        at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:358)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:311)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)
        at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1656)
        at org.apache.gravitino.server.authentication.AuthenticationFilter.doFilter(AuthenticationFilter.java:73)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626)
        at org.apache.gravitino.server.web.VersioningFilter.doFilter(VersioningFilter.java:111)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:552)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:600)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.Server.handle(Server.java:516)
        at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)
        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
        at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:137)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
        at java.lang.Thread.run(Thread.java:750)
        at org.apache.gravitino.client.ErrorHandlers$CatalogErrorHandler.accept(ErrorHandlers.java:355)
        at org.apache.gravitino.client.ErrorHandlers$CatalogErrorHandler.accept(ErrorHandlers.java:345)
        at org.apache.gravitino.client.HTTPClient.throwFailure(HTTPClient.java:225)
        at org.apache.gravitino.client.HTTPClient.execute(HTTPClient.java:378)
        at org.apache.gravitino.client.HTTPClient.execute(HTTPClient.java:286)
        at org.apache.gravitino.client.HTTPClient.post(HTTPClient.java:480)
        at org.apache.gravitino.client.GravitinoMetalake.createCatalog(GravitinoMetalake.java:171)
        at org.apache.gravitino.flink.connector.integration.test.hive.FlinkHiveCatalogIT.initDefaultHiveCatalog(FlinkHiveCatalogIT.java:90)
        at org.apache.gravitino.flink.connector.integration.test.hive.FlinkHiveCatalogIT.hiveStartUp(FlinkHiveCatalogIT.java:78)
```

**Comment by coolderli** [comment]
@FANNG1 Please help approve the workflow and review this MR. Thanks.

File: flink-connector/build.gradle.kts (hunk 1 above)
Comment (FANNG1) [reply flink-connector/build.gradle.kts]: maybe you should use
```
tasks.test {
    dependsOn(":catalogs:catalog-hive:jar")
}
```

**Comment by coolderli** [comment]
@FANNG1 Please help review. Thanks.

**Comment by FANNG1** [comment]
@mchades do you have time to review this PR?

File: flink-connector/src/main/java/org/apache/gravitino/flink/connector/utils/TypeUtils.java [hunk 2]
```diff
@@ -39,6 +47,69 @@ public static Type toGravitinoType(LogicalType logicalType) {
         return Types.IntegerType.get();
       case BIGINT:
         return Types.LongType.get();
+      case CHAR:
+        CharType charType = (CharType) logicalType;
+        return Types.FixedCharType.of(charType.getLength());
+      case BOOLEAN:
+        return Types.BooleanType.get();
+      case BINARY:
```
Comment (FANNG1) [inline flink-connector/src/main/java/org/apache/gravitino/flink/connector/utils/TypeUtils.java:55]: flink `BINARY` has fixed length, seems couldn't convert to Gravitino?

File: flink-connector/src/main/java/org/apache/gravitino/flink/connector/utils/TypeUtils.java [hunk 3]
```diff
@@ -39,6 +47,69 @@ public static Type toGravitinoType(LogicalType logicalType) {
...
+      case FLOAT:
+        return Types.FloatType.get();
+      case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+      case TIMESTAMP_WITH_TIME_ZONE:
+        return Types.TimestampType.withTimeZone();
+      case ARRAY:
+        ArrayType arrayType = (ArrayType) logicalType;
+        Type elementType = toGravitinoType(arrayType.getElementType());
+        return Types.ListType.of(elementType, arrayType.isNullable());
+      case MULTISET:
```
Comment (FANNG1) [inline flink-connector/src/main/java/org/apache/gravitino/flink/connector/utils/TypeUtils.java:84]: could `MULTISET`  transformed to `Map`?

File: flink-connector/src/main/java/org/apache/gravitino/flink/connector/utils/TypeUtils.java (hunk 2 above)
Comment (FANNG1) [reply flink-connector/src/main/java/org/apache/gravitino/flink/connector/utils/TypeUtils.java:55]: `BINARY` could transform to `FIXED`

**Comment by coolderli** [comment]
flink type doc:
https://nightlies.apache.org/flink/flink-docs-master/zh/docs/dev/table/types/

**Comment by coolderli** [comment]
@FANNG1 Please take a look. Thanks.

File: flink-connector/src/main/java/org/apache/gravitino/flink/connector/utils/TypeUtils.java [hunk 4]
```diff
@@ -55,8 +129,80 @@ public static DataType toFlinkType(Type gravitinoType) {
...
+        return DataTypes.SMALLINT();
+      case DECIMAL:
+        Types.DecimalType decimalType = (Types.DecimalType) gravitinoType;
+        return DataTypes.DECIMAL(decimalType.precision(), decimalType.scale());
+      case VARCHAR:
+        Types.VarCharType varCharType = (Types.VarCharType) gravitinoType;
+        return DataTypes.VARCHAR(varCharType.length());
+      case FIXED:
+        Types.FixedType fixedType = (Types.FixedType) gravitinoType;
+        return DataTypes.CHAR(fixedType.length());
```
Comment (FANNG1) [inline flink-connector/src/main/java/org/apache/gravitino/flink/connector/utils/TypeUtils.java:144]: fixed -> binary?

**Comment by FANNG1** [comment]
LGTM except one comment

**Comment by coolderli** [comment]
@FANNG1 Please take a look. Thanks.

**Pull request merged**